lib/code-caller-docker.js

```diff
@@ -0,0 +1,743 @@
+    this.debug(`creating bind mount for ${directory} at ${mountpoint}`);
+    await createBindMount(directory, mountpoint);
+    this.debug(`created bind mount for ${directory} at ${mountpoint}`);
+  }
+
+  /**
+   * Wrapper around `removeBindMount` that includes instance-specific logs.
+   *
+   * @param {string} mountpoint
+   */
+  async removeBindMount(mountpoint) {
+    this.debug(`removing bind mount at ${mountpoint}`);
+    await removeBindMount(mountpoint);
+    this.debug(`removed bind mount at ${mountpoint}`);
+  }
+
+  /**
+   * Allows this caller to prepare for execution of code from a particular
+   * course.
+   *
+   * @param {string} coursePath
+   * @param {(err: Error | null | undefined) => void} callback
+   */
+  prepareForCourse(coursePath, callback) {
+    util.callbackify(async () => {
+      if (this.coursePath && this.coursePath === coursePath) {
+        // Same course as before; we can reuse the existing setup
+        return;
+      }
+
+      this.coursePath = coursePath;
+
+      if (config.workersExecutionMode === 'container') {
+        await this._ensureDirectoryIfNeeded();
+        // If this Docker code caller was used before, we might have an existing
+        // bind mount that we need to remove first. This should have been cleaned
+        // up before, but if it wasn't, we'll do so here to avoid a
+        // `mount point XXX is itself on a OSXFUSE volume` error.
+        try {
+          await this.removeBindMount(this.hostDirectory.path);
+        } catch (e) {
+          // If the above command failed, assume it because a mount point didn't
+          // already exist and proceed as normal. If something is really messed
+          // up, the next command will likely fail anyways.
+        }
+        await this.createBindMount(coursePath, this.hostDirectory.path);
+      } else {
+        throw new Error(`Unexpected worker execution mode "${config.workersExecutionMode}"`);
+      }
+    })(callback);
+  }
+
+  /**
+   *
+   * @param {'restart' | 'question' | 'course-element' | 'core-element'} type
+   * @param {string} directory
+   * @param {string} file
+   * @param {string} fcn
+   * @param {string[]} args
+   * @param {(err: Error | null | undefined, data?: any, output?: string) => void} callback
+   */
+  call(type, directory, file, fcn, args, callback) {
+    this.debug('enter call()');
+    this.callCount += 1;
+    this.ensureChild()
+      .then(() => {
+        if (!this._checkState([WAITING])) return callback(new Error('invalid DockerCaller state'));
+        if (!this._checkReadyForCall()) return callback(new Error('not ready for call'));
+        // TODO: state checking might be off. for the 'course' allocation method, we only start
+        // containers when we know about the course. however, for the general case in prod, we can
+        // start containers immediately on boot by "warming up". Need to make sure we actually do that,
+        // and that it's handled correctly.
```

Comment:
I believe this comment is obsolete - `lib/code-callers` should always warm up during boot as long as we aren't configured not to, and we don't have any notion of multiple allocation strategies anymore.

Reply:
Done in 9fe5e70.